code/modules/mining/laborcamp/laborshuttle.dm

```diff
@@ -6,10 +6,180 @@
+		ui.open()
+
+/obj/machinery/mineral/labor_prisoner_shuttle_console/ui_data(mob/user)
+	var/list/data = list()
+	var/can_go_home = FALSE
+
+	data["emagged"] = emagged
+	data["id_inserted"] = inserted_id != null
+	if(inserted_id)
+		data["id_name"] = inserted_id.registered_name
+		data["id_points"] = inserted_id.mining_points
+		data["id_goal"] = inserted_id.goal
+
+	if(check_auth())
+		can_go_home = TRUE
+
+	data["can_go_home"] = can_go_home
+
+	return data
+
+/obj/machinery/mineral/labor_prisoner_shuttle_console/ui_act(action, params)
+	if(..())
+		return
+
+	switch(action)
+		if("handle_id")
+			if(inserted_id)
+				if(!usr.put_in_hands(inserted_id))
+					inserted_id.forceMove(get_turf(src))
+				inserted_id = null
+			else
+				var/obj/item/I = usr.get_active_hand()
+				if(istype(I, /obj/item/card/id/prisoner))
+					if(!usr.unEquip(I))
+						return
+					I.forceMove(src)
+					inserted_id = I
+		if("move_shuttle")
+			if(!alone_in_area(get_area(src), usr))
+				to_chat(usr, "<span class='warning'>Prisoners are only allowed to be released while alone.</span>")
+			else
+				switch(SSshuttle.moveShuttle("laborcamp", "laborcamp_home", TRUE, usr))
+					if(1)
+						to_chat(usr, "<span class='notice'>Shuttle not found.</span>")
+					if(2)
+						to_chat(usr, "<span class='notice'>Shuttle already at station.</span>")
+					if(3)
+						to_chat(usr, "<span class='notice'>No permission to dock could be granted.</span>")
+					else
+						if(!emagged)
+							var/message = "[inserted_id.registered_name] has returned to the station. Minerals and Prisoner ID card ready for retrieval."
+							announcer.autosay(message, "Labor Camp Controller", "Security")
+							sec_record_release(usr)
+						to_chat(usr, "<span class='notice'>Shuttle received message and will be sent shortly.</span>")
+						usr.create_log(MISC_LOG, "used [src] to call the laborcamp shuttle")
+
+	return TRUE
+
+/obj/machinery/mineral/labor_prisoner_shuttle_console/proc/sec_record_release(mob/to_be_released)
+	var/mob/living/carbon/human/target = to_be_released
+	var/target_name = target.get_visible_name(TRUE)
+	if(!target_name || target_name == "Unknown")
+		return //Do you have the slightest idea how little that narrows it down?
+	var/datum/data/record/R = find_record("name", target_name, GLOB.data_core.security)
+	if(!(R?.fields["criminal"] in list(SEC_RECORD_STATUS_INCARCERATED)))
+		return //stops cheese
+	set_criminal_status(to_be_released, R, SEC_RECORD_STATUS_RELEASED, "Automatically released upon reaching labor points quota", "Labor Camp Controller", user_name = "Automated System")
+
+/obj/machinery/mineral/labor_prisoner_shuttle_console/proc/check_auth()
+	if(emagged)
+		return TRUE //Shuttle is emagged, let any ol' person through
+	if(!inserted_id?.goal)
+		return FALSE //ID not properly set up, abort!
+	return (istype(inserted_id) && inserted_id.mining_points >= inserted_id.goal) //Otherwise, only let them out if the prisoner's reached his quota.
```

Comment:
what is the point of the istype check here? `inserted_id` should always be the type of the var or else someone wrote bad code somewhere, besides you already try and access the `goal` var above so if it's not the correct type it will runtime.

Reply:
No idea, most of this code is moved from another file I'm not the author. I suppose I should have anticipated becoming responsible for it and cleaned it up.

Reply:
💀  Yeah, using old code/refactoring it is painful in that sense.